src/Microsoft.AspNetCore.Mvc.Core/Internal/ApiBehaviorApplicationModelProvider.cs

```diff
@@ -0,0 +1,177 @@
+
+        public ApiBehaviorApplicationModelProvider(
+            IOptions<ApiBehaviorOptions> apiBehaviorOptions,
+            IModelMetadataProvider modelMetadataProvider,
+            ILoggerFactory loggerFactory)
+        {
+            _apiBehaviorOptions = apiBehaviorOptions.Value;
+            _modelMetadataProvider = modelMetadataProvider as ModelMetadataProvider;
+            _logger = loggerFactory.CreateLogger<ApiBehaviorApplicationModelProvider>();
+
+            if (_apiBehaviorOptions.EnableModelStateInvalidFilter && _apiBehaviorOptions.InvalidModelStateResponseFactory == null)
+            {
+                throw new ArgumentException(Resources.FormatPropertyOfTypeCannotBeNull(
+                    typeof(ApiBehaviorOptions),
+                    nameof(ApiBehaviorOptions.InvalidModelStateResponseFactory)));
+            }
+
+            _modelStateInvalidFilter = new ModelStateInvalidFilter(
+                apiBehaviorOptions.Value,
+                loggerFactory.CreateLogger<ModelStateInvalidFilter>());
+        }
+
+        /// <remarks>
+        /// Order is set to execute after the <see cref="DefaultApplicationModelProvider"/> and allow any other user
+        /// <see cref="IApplicationModelProvider"/> that configure routing to execute.
+        /// </remarks>
+        public int Order => -1000 + 100;
+
+        public void OnProvidersExecuted(ApplicationModelProviderContext context)
+        {
+        }
+
+        public void OnProvidersExecuting(ApplicationModelProviderContext context)
+        {
+            foreach (var controllerModel in context.Result.Controllers)
+            {
+                var isApiController = controllerModel.Attributes.OfType<IApiBehaviorMetadata>().Any();
+                var controllerHasSelectorModel = controllerModel.Selectors.Any(s => s.AttributeRouteModel != null);
+
+                foreach (var actionModel in controllerModel.Actions)
+                {
+                    if (!isApiController && !actionModel.Attributes.OfType<IApiBehaviorMetadata>().Any())
+                    {
+                        continue;
+                    }
+
+                    if (!controllerHasSelectorModel && !actionModel.Selectors.Any(s => s.AttributeRouteModel != null))
+                    {
+                        // Require attribute routing with controllers annotated with ApiControllerAttribute
+                        throw new InvalidOperationException(Resources.FormatApiController_AttributeRouteRequired(nameof(ApiControllerAttribute)));
+                    }
+
+                    if (_apiBehaviorOptions.EnableModelStateInvalidFilter)
+                    {
+                        Debug.Assert(_apiBehaviorOptions.InvalidModelStateResponseFactory != null);
+                        actionModel.Filters.Add(_modelStateInvalidFilter);
+                    }
+
+                    if (_modelMetadataProvider != null && _apiBehaviorOptions.InferBindingSourcesForParameters)
```

Comment:
What about bound properties? IMO we should do the same thing

Reply:
Isn't that problematic? We'd be looking at properties without any BindingInfo and assigning them binding sources. Plus we'd have to know about other things that are ModelBinding-ish (like `TempData`) and make we're not attempting to model bind them.